Revise matrix factorization (#47)
greenelab/knowledge-graph-review: content/04.network_applications.md

$A$ is constructed based on the knowledge graph's structure itself and collapses all edges between two nodes into one unique entity.
Following construction, $A$ is decomposed into the following three parts: a square matrix $Σ$ and a set of two small rectangular matrices $U$ and $V^{T}$.
Values within $Σ$ are called singular values, which are akin to eigenvalues [@doi:10.1007/BF02288367].
Each row in $U$ and each column in $V^{T}$ represents nodes within a low dimensional space [@doi:10.1007/BF02288367; @doi:10/bm8dnf].
In practice $U$ is usually used to represent nodes in a knowledge graph and can be used as input for machine learning classifiers to perform tasks such as link prediction or node classification [@doi:10.1093/bioinformatics/btz718]; however,$V^{T}$ can also be used [@doi:10.1007/BF02288367; @raw:Jiezhon2018].
Typically, matrix factorization algorithms such as SVD appear in recommendation systems via collaborative filtering [@doi:10.1155/2009/421425]; however, this technique can also be used as a standalone baseline to compare to other relational learning approaches [@doi:10.1093/bioinformatics/btz718].

> Laplacian eigenmaps is a technique that assumes there is low dimensional structure in a high dimensional space and preserves this structure when projecting data into a low dimensional space [@doi:10.1162/089976603321780317].

Reviewer: ```suggestion
Laplacian eigenmaps assume there is low dimensional structure in a high dimensional space and preserves this structure when projecting data into a low dimensional space [@doi:10.1162/089976603321780317].
```

> Typically, the first step of this technique is to construct a figurative graph where nodes represent datapoints and edges are constructed based on the similarity of two datapoints; however, in the context of knowledge graphs the graph has already been defined.

Reviewer: By this do you mean that the knowledge graph is the graph that one needs? If so, maybe there's a way to make that clearer.

These methods were designed to be used on many different types of data; however, we discuss their use in the context of representing knowledge graphs in a low dimensional space and focus particularly on SVD and laplacian eigenmaps.  

SVD [@doi:10.1007/BF02288367] is an algorithm that uses matrix factorization to portray knowledge graphs in a low dimensional space.
The input for this algorithm is an adjacency matrix ($A$), which is a square matrix where rows and columns represent nodes and each entry is a binary representation of the presence of an edge between two nodes. 
$A$ is constructed based on the knowledge graph's structure itself and collapses all edges between two nodes into one unique entity.
Following construction, $A$ is decomposed into the following three parts: a square matrix $Σ$ and a set of two small rectangular matrices $U$ and $V^{T}$.
Values within $Σ$ are called singular values, which are akin to eigenvalues [@doi:10.1007/BF02288367].
Each row in $U$ and each column in $V^{T}$ represents nodes within a low dimensional space [@doi:10.1007/BF02288367; @doi:10/bm8dnf].

> In practice $U$ is usually used to represent nodes in a knowledge graph and can be used as input for machine learning classifiers to perform tasks such as link prediction or node classification [@doi:10.1093/bioinformatics/btz718]; however,$V^{T}$ can also be used [@doi:10.1007/BF02288367; @raw:Jiezhon2018].

Reviewer: ```suggestion
In practice $U$ is usually used to represent nodes in a knowledge graph and can be used as input for machine learning classifiers to perform tasks such as link prediction or node classification [@doi:10.1093/bioinformatics/btz718]; however,$V^{T}$ has also been used [@doi:10.1007/BF02288367; @raw:Jiezhon2018].
```

> Typically, matrix factorization algorithms such as SVD appear in recommendation systems via collaborative filtering [@doi:10.1155/2009/421425]; however, this technique can also be used as a standalone baseline to compare to other relational learning approaches [@doi:10.1093/bioinformatics/btz718].

Reviewer: ```suggestion
Typically, matrix factorization algorithms such as SVD are used for recommendation systems via collaborative filtering [@doi:10.1155/2009/421425]; however, this technique can also provide a standalone baseline for other relational learning approaches [@doi:10.1093/bioinformatics/btz718].
```

Laplacian eigenmaps assume there is low dimensional structure in a high dimensional space and preserves this structure when projecting data into a low dimensional space [@doi:10.1162/089976603321780317].
The first step of this technique is to preserve this low dimensional structure by representing data in the form of a graph where nodes are datapoints and edges are the distance between two points.

> In the context of knowledge graphs the graph representation is already defined, which means one can move on towards the next step.

Reviewer: ```suggestion
Knowledge graphs already provide this representation, so no additional processing is necessary at this stage.
```

> The second step of this technique is to obtain both an adjacency matrix ($A$) and a degree matrix ($D$) from graph representation.

Reviewer: ```suggestion
The second step of this technique is to obtain both an adjacency matrix ($A$) and a degree matrix ($D$) from the graph representation.
```

Typically, eigenmaps work well when knowledge graphs have a sparse number of edges between nodes but struggle when presented with denser networks [@doi:10.1093/bioinformatics/btz718; @doi:10.1371/journal.pcbi.1005621; @arxiv:1905.09763].
An open area of exploration is to adapt these methods to accommodate knowledge graphs that have a large number of edges.

> Matrix factorization is a powerful technique that uses matrices such as an adjacency matrix as input.

Reviewer: Is there a context in which you'd use this where you wouldn't be factorizing the adjacency matrix? I think you could simplify this sentence if not.

Typically, eigenmaps work well when knowledge graphs have a sparse number of edges between nodes but struggle when presented with denser networks [@doi:10.1093/bioinformatics/btz718; @doi:10.1371/journal.pcbi.1005621; @arxiv:1905.09763].
An open area of exploration is to adapt these methods to accommodate knowledge graphs that have a large number of edges.

Matrix factorization is a powerful technique that represents high dimensional data in a low dimensional space.
Common approaches involve using SVD, Laplacian eigenmaps or variants of the two to decompose matrices into smaller rectangular forms.
Regarding knowledge graphs, the adjacency matrix ($A$) is the typical matrix that gets decomposed, but the laplacian matrix ($L=D-A$) can be used as well. 

> Despite reported success, the dependence on matrices creates an issue of scalability as matrices of large networks would take too much memory for a regular computer to operate well.

Reviewer: ```suggestion
Despite reported success, the dependence on matrices creates an issue of scalability as matrices of large networks may reach memory limitations.
```

> Furthermore, these methods treat all edge types the same, but a possible enhancement is to incorporate node and edge types as sources of input.

Reviewer: ```suggestion
Furthermore, the approaches we discussed consider all edge types as equivalent.
This limitations could be mitigated by new approaches designed to accommodate multiple node and edge types separately.
```
